numpy/fft/_pocketfft.py

```diff
@@ -46,33 +46,36 @@
 # divided. This replaces the original, more intuitive 'fct` parameter to avoid
 # divisions by zero (or alternatively additional checks) in the case of
 # zero-length axes during its computation.
-def _raw_fft(a, n, axis, is_real, is_forward, inv_norm):
+def _raw_fft(a, n, axis, is_real, is_forward, inv_norm, out=None):
     axis = normalize_axis_index(axis, a.ndim)
     if n is None:
         n = a.shape[axis]
 
     fct = 1/inv_norm
 
-    if a.shape[axis] != n:
-        s = list(a.shape)
-        index = [slice(None)]*len(s)
-        if s[axis] > n:
-            index[axis] = slice(0, n)
-            a = a[tuple(index)]
+    n_out = n
+    if is_real:
+        if is_forward:
+            ufunc = pfu.rfft_n_even if n % 2 == 0 else pfu.rfft_n_odd
+            n_out = n // 2 + 1
         else:
-            index[axis] = slice(0, s[axis])
-            s[axis] = n
-            z = zeros(s, a.dtype.char)
-            z[tuple(index)] = a
-            a = z
-
-    if axis == a.ndim-1:
-        r = pfi.execute(a, is_real, is_forward, fct)
+            ufunc = pfu.irfft
     else:
-        a = swapaxes(a, axis, -1)
-        r = pfi.execute(a, is_real, is_forward, fct)
-        r = swapaxes(r, axis, -1)
-    return r
+        ufunc = pfu.fft if is_forward else pfu.ifft
+
+    if out is None:
+        out = empty(a.shape[:axis] + (n_out,) + a.shape[axis+1:],
+                    dtype=complex if is_forward or not is_real else float)
+    elif ((shape := getattr(out, "shape", None)) is not None
```

Comment:
Really this is there just because I need to do the check on the size of the output axis, which I cannot leave to the ufunc machinery since it does not have to be the same as the input size. But, indeed, to check the `shape` I have to be sure it exists...